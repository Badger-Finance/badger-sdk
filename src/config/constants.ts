import { ARB_CONFIG } from './network/arbitrum.config';
import { AVAX_CONFIG } from './network/avalanche.config';
import { BSC_CONFIG } from './network/binance.config';
import { ETH_CONFIG } from './network/ethereum.config';
import { FTM_CONFIG } from './network/fantom.config';
import { LOCAL_CONFIG } from './network/local.config';
import { MATIC_CONFIG } from './network/polygon.config';

export const ONE_MIN_MS = 60 * 1000;
export const ONE_HOUR_MS = 60 * ONE_MIN_MS;
export const ONE_DAY_MS = 24 * ONE_HOUR_MS;
export const ONE_YEAR_MS = 365 * ONE_DAY_MS;

export const SUPPORTED_NETWORKS = [
  ETH_CONFIG,
  LOCAL_CONFIG,
  BSC_CONFIG,
  MATIC_CONFIG,
  ARB_CONFIG,
  AVAX_CONFIG,
  FTM_CONFIG,
];

export const VERSION = '1.0.77';
