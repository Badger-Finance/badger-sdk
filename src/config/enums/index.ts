export { Network } from './network.enum';
