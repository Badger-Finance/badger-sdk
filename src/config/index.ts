export * from './enums';
export * from './types';
export * from './constants';
export * from './interfaces';
export * from './network';
