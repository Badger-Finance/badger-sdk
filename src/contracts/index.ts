/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
export type { BadgerTree } from "./BadgerTree";
export type { Digg } from "./Digg";
export type { Erc20 } from "./Erc20";
export type { Ibbtc } from "./Ibbtc";
export type { Registry } from "./Registry";
export type { RewardsLogger } from "./RewardsLogger";
export type { Sett } from "./Sett";
export type { StrategyV15 } from "./StrategyV15";
export type { VaultV15 } from "./VaultV15";

export { BadgerTree__factory } from "./factories/BadgerTree__factory";
export { Digg__factory } from "./factories/Digg__factory";
export { Erc20__factory } from "./factories/Erc20__factory";
export { Ibbtc__factory } from "./factories/Ibbtc__factory";
export { Registry__factory } from "./factories/Registry__factory";
export { RewardsLogger__factory } from "./factories/RewardsLogger__factory";
export { Sett__factory } from "./factories/Sett__factory";
export { StrategyV15__factory } from "./factories/StrategyV15__factory";
export { VaultV15__factory } from "./factories/VaultV15__factory";
