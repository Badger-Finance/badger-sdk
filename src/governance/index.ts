export { GovernanceService } from './governance.service';
