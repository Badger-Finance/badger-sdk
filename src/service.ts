import { providers } from '@0xsequence/multicall';
import { Signer } from 'ethers';

import { BadgerSDK, NetworkConfig } from '.';
import BadgerAPI from './api';
import { Logger } from './logger';

export abstract class Service extends Logger {
  constructor(protected sdk: BadgerSDK) {
    super(sdk.logLevel);
  }

  get provider(): providers.MulticallProvider {
    return this.sdk.provider;
  }

  get signer(): Signer | undefined {
    return this.sdk.signer;
  }

  get connector(): providers.MulticallProvider | Signer {
    return this.address && this.signer ? this.signer : this.provider;
  }

  get config(): NetworkConfig {
    return this.sdk.config;
  }

  get address(): string | undefined {
    return this.sdk.address;
  }

  get api(): BadgerAPI {
    return this.sdk.api;
  }
}
