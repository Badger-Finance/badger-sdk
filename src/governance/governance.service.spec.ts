import { mock, MockProxy } from 'jest-mock-extended';

import { Network } from '../config';
import { TimelockController, TimelockController__factory } from '../contracts';
import { BadgerSDK } from '../sdk';
import { mockSDK } from '../tests';
import { TestServiceEnum } from '../tests/service.enum';
import * as deployedAtUtils from '../utils/deployed-at.util';
import { GovernanceService } from './governance.service';
import { GovernanceProxyMock } from './mocks/governance.proxy.mock';
import { PROPOSALS_CANCELLED_EVENTS } from './mocks/proposal.cancelled.event';
import { PROPOSALS_EXECUTED_EVENTS } from './mocks/proposal.executed.event';
import { PROPOSALS_DISPUTED_EVENTS } from './mocks/proposals.disputed.event';
import { PROPOSALS_RESOLVED_EVENTS } from './mocks/proposals.resolved.event';
import { PROPOSALS_SCHEDULED_EVENTS } from './mocks/proposals.scheduled.event';

describe('governance.service', () => {
  let sdk: BadgerSDK;
  let timeLockControllerMock: MockProxy<TimelockController>;

  beforeEach(async function () {
    sdk = mockSDK(Network.Arbitrum, TestServiceEnum.governance);

    const timelockController = mock<TimelockController>();

    jest
      .spyOn(TimelockController__factory, 'connect')
      .mockImplementation(() => timelockController);

    timeLockControllerMock = timelockController;

    await sdk.ready();
  });

  describe('timelock controller address getter', () => {
    it('should return the timeLock address', async function () {
      expect(
        Object.keys(GovernanceService.TIMELOCK_CONTRACT_ADDR_MAP).length,
      ).toBeGreaterThan(0);
      expect(
        Object.values(GovernanceService.TIMELOCK_CONTRACT_ADDR_MAP).length,
      ).toBeGreaterThan(0);
      expect(sdk.governance.timelockAddress).toMatchSnapshot();
    });

    it('should be undefined if the network is not supported', async function () {
      const tempTimelockAddressMap = {
        ...GovernanceService.TIMELOCK_CONTRACT_ADDR_MAP,
      };

      Object.defineProperty(GovernanceService, 'TIMELOCK_CONTRACT_ADDR_MAP', {
        value: {},
      });

      expect(sdk.governance.timelockAddress).toBeUndefined();

      Object.defineProperty(GovernanceService, 'TIMELOCK_CONTRACT_ADDR_MAP', {
        value: tempTimelockAddressMap,
      });
    });
  });

  describe('timelock contract getter', () => {
    it('should return the timelock contract', async () => {
      expect(sdk.governance.timelockController).toBe(timeLockControllerMock);
    });

    it('should throw an error if it not deployed on the network', () => {
      const tempTimelockAddressMap = {
        ...GovernanceService.TIMELOCK_CONTRACT_ADDR_MAP,
      };
      sdk = mockSDK(Network.Fantom, TestServiceEnum.governance);

      Object.defineProperty(GovernanceService, 'TIMELOCK_CONTRACT_ADDR_MAP', {
        value: {},
      });

      expect(() => sdk.governance.timelockController).toThrowError();

      Object.defineProperty(GovernanceService, 'TIMELOCK_CONTRACT_ADDR_MAP', {
        value: tempTimelockAddressMap,
      });
    });
  });

  describe('processEventsScanRange', () => {
    let governance: GovernanceProxyMock;

    beforeEach(() => {
      governance = new GovernanceProxyMock(sdk);
    });

    it('should return options values if they are defined', async () => {
      const options = {
        startBlock: 1,
        endBlock: 2,
      };

      await governance.processEventsScanRangePr(
        governance.timelockAddress,
        options,
      );

      expect(options.startBlock).toBe(1);
      expect(options.endBlock).toBe(2);
    });

    it('should assign the block range if they are not defined', async () => {
      const startBlock = 1000;
      const endBlock = 999999;

      jest
        .spyOn(deployedAtUtils, 'getBlockDeployedAt')
        .mockReturnValue(startBlock);
      jest.spyOn(sdk.provider, 'getBlockNumber').mockResolvedValue(endBlock);

      const options = {
        startBlock: 0,
        endBlock: 0,
      };

      await governance.processEventsScanRangePr(
        governance.timelockAddress,
        options,
      );

      expect(options.startBlock).toBe(startBlock);
      expect(options.endBlock).toBe(endBlock);
    });
  });

  describe('loadScheduledProposals', () => {
    beforeEach(() => {
      timeLockControllerMock.queryFilter.mockImplementation(
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        async (_, startBlock, endBlock) => {
          return PROPOSALS_SCHEDULED_EVENTS.filter(
            (event) =>
              event.blockNumber >= Number(startBlock) &&
              event.blockNumber <= Number(endBlock),
          );
        },
      );
      timeLockControllerMock.filters.CallScheduled = jest.fn();
    });

    it('should return scheduled proposals', async () => {
      const proposals = await sdk.governance.loadScheduledProposals({
        startBlock: 34_941_486,
        endBlock: 34_941_490,
      });

      expect(proposals).toMatchSnapshot();
    });

    it('should return empty array if there are no scheduled proposals within blocks range', async () => {
      const proposals = await sdk.governance.loadScheduledProposals({
        startBlock: 10,
        endBlock: 100,
      });

      expect(proposals).toEqual([]);
    });
  });

  describe('loadProposalsStatusChanges', () => {
    const defaultRangeOpts = {
      startBlock: 34_941_480,
      endBlock: 34_941_490,
    };

    const eventCallExecuted = 'CallExecuted';
    const eventCallCanceled = 'Canceled';

    beforeEach(() => {
      timeLockControllerMock.queryFilter.mockImplementation(
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        async (ef, startBlock, endBlock) => {
          const filterFunc = <T extends Record<string, number>>(event: T) =>
            event.blockNumber >= Number(startBlock) &&
            event.blockNumber <= Number(endBlock);

          if (ef.topics?.includes(eventCallExecuted)) {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            return PROPOSALS_EXECUTED_EVENTS.filter(filterFunc);
          } else if (ef.topics?.includes(eventCallCanceled)) {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            return PROPOSALS_CANCELLED_EVENTS.filter(filterFunc);
          }

          return [];
        },
      );
      timeLockControllerMock.filters.CallExecuted = jest.fn().mockReturnValue({
        topics: [eventCallExecuted],
      });
      timeLockControllerMock.filters.Cancelled = jest.fn().mockReturnValue({
        topics: [eventCallCanceled],
      });
    });

    it('should return events with status changes for proposals in sorted way', async () => {
      const events = await sdk.governance.loadProposalsStatusChange(
        defaultRangeOpts,
      );

      expect(events).toMatchSnapshot();
    });

    it('should return empty array if there are no scheduled proposals within blocks range', async () => {
      const events = await sdk.governance.loadProposalsStatusChange({
        startBlock: 10,
        endBlock: 100,
      });

      expect(events).toEqual([]);
    });
  });

  describe('loadProposalsDisputes', () => {
    const defaultRangeOpts = {
      startBlock: 34_941_480,
      endBlock: 34_941_490,
    };

    const eventCallDisputed = 'CallDisputed';
    const eventDisputedResolved = 'CallDisputedResolved';

    beforeEach(() => {
      timeLockControllerMock.queryFilter.mockImplementation(
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        async (ef, startBlock, endBlock) => {
          const filterFunc = <T extends Record<string, number>>(event: T) =>
            event.blockNumber >= Number(startBlock) &&
            event.blockNumber <= Number(endBlock);

          if (ef.topics?.includes(eventCallDisputed)) {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            return PROPOSALS_DISPUTED_EVENTS.filter(filterFunc);
          } else if (ef.topics?.includes(eventDisputedResolved)) {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            return PROPOSALS_RESOLVED_EVENTS.filter(filterFunc);
          }

          return [];
        },
      );
      timeLockControllerMock.filters.CallDisputed = jest.fn().mockReturnValue({
        topics: [eventCallDisputed],
      });
      timeLockControllerMock.filters.CallDisputedResolved = jest
        .fn()
        .mockReturnValue({
          topics: [eventDisputedResolved],
        });
    });

    it('should return events with status changes for proposals in sorted way', async () => {
      const events = await sdk.governance.loadProposalsDisputes(
        defaultRangeOpts,
      );

      expect(events).toMatchSnapshot();
    });

    it('should return empty array if there are no scheduled proposals within blocks range', async () => {
      const events = await sdk.governance.loadProposalsDisputes({
        startBlock: 10,
        endBlock: 100,
      });

      expect(events).toEqual([]);
    });
  });
});
