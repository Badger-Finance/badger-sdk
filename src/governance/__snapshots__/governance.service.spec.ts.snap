// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`governance.service loadProposalsDisputes should return events with status changes for proposals 1`] = `
Array [
  Object {
    "address": "0xC4611233328559Dd1849947bC1e2b3B6b8714fa9",
    "args": Array [
      "0xe9bfab585a583ae5cc2395f40bfb61b5c93dee4356079f1c642d7157be8300c8",
      "0xef42D748e09A2d9eF89238c053CE0B6f00236210",
      "Disputed",
    ],
    "blockHash": "0xb48bea6fe4f87cd4323437a9700529db1a8ab5f0376835e46a403f5dfe16806f",
    "blockNumber": 34941480,
    "data": "0x000000000000000000000000fc13209cafe8fb3bb5fbd929ec9f11a39e8ac041000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000638bc098000000000000000000000000ef42d748e09a2d9ef89238c053ce0b6f00236210000000000000000000000000000000000000000000000000000000000000014000000000000000000000000000000000000000000000000000000000000000248a0dac4a000000000000000000000000c69fb085481bc8c4bff99b924076656305d9a25d00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000850726f706f736564000000000000000000000000000000000000000000000000",
    "event": "CallDisputed",
    "eventSignature": "CallDisputed(bytes32,address,string)",
    "logIndex": 6,
    "removed": false,
    "topics": Array [
      "0x21f5c7d6a15760b2a1e87a9b8cff0ad5e6423cd8f41e04f2196934b1d3859de6",
      "0xe9bfab585a583ae5cc2395f40bfb61b5c93dee4356079f1c642d7157be8300c8",
      "0x0000000000000000000000000000000000000000000000000000000000000000",
    ],
    "transactionHash": "0xdb286638e5dc3a62c2c814fe61016ad0e49349bbba0fcce2b36640a42b70cac3",
    "transactionIndex": 2,
  },
  Object {
    "address": "0xC4611233328559Dd1849947bC1e2b3B6b8714fa9",
    "args": Array [
      "0xe9bfab585a583ae5cc2395f40bfb61b5c93dee4356079f1c642d7157be8300c8",
      "0xef42D748e09A2d9eF89238c053CE0B6f00236210",
      "Disputed",
    ],
    "blockHash": "0xb48bea6fe4f87cd4323437a9700529db1a8ab5f0376835e46a403f5dfe16806f",
    "blockNumber": 34941481,
    "data": "0x000000000000000000000000e774d1fb3133b037aa17d39165b8f45f444f632d000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000638bc098000000000000000000000000ef42d748e09a2d9ef89238c053ce0b6f0023621000000000000000000000000000000000000000000000000000000000000001400000000000000000000000000000000000000000000000000000000000000024c7b9d530000000000000000000000000e8ea1d8b3a5a4cec7e94ae330ff18e82b5d22fa600000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000850726f706f736564000000000000000000000000000000000000000000000000",
    "event": "CallDisputed",
    "eventSignature": "CallDisputed(bytes32,address,string)",
    "logIndex": 7,
    "removed": false,
    "topics": Array [
      "0x21f5c7d6a15760b2a1e87a9b8cff0ad5e6423cd8f41e04f2196934b1d3859de6",
      "0xe9bfab585a583ae5cc2395f40bfb61b5c93dee4356079f1c642d7157be8300c8",
      "0x0000000000000000000000000000000000000000000000000000000000000001",
    ],
    "transactionHash": "0xdb286638e5dc3a62c2c814fe61016ad0e49349bbba0fcce2b36640a42b70cac3",
    "transactionIndex": 2,
  },
]
`;

exports[`governance.service loadProposalsStatusChanges should return events with status changes for proposals in sorted way 1`] = `
Array [
  Object {
    "address": "0xC4611233328559Dd1849947bC1e2b3B6b8714fa9",
    "args": Array [
      "0xe9bfab585a583ae5cc2395f40bfb61b5c93dee4356079f1c642d7157be8300c8",
      Object {
        "hex": "0x00",
        "type": "BigNumber",
      },
      "0xFc13209cAfE8fb3bb5fbD929eC9F11a39e8Ac041",
      Object {
        "hex": "0x00",
        "type": "BigNumber",
      },
      "0x8a0dac4a000000000000000000000000c69fb085481bc8c4bff99b924076656305d9a25d",
      "0x0000000000000000000000000000000000000000000000000000000000000000",
      "Executed",
    ],
    "blockHash": "0xb48bea6fe4f87cd4323437a9700529db1a8ab5f0376835e46a403f5dfe16806f",
    "blockNumber": 34941480,
    "data": "0x000000000000000000000000fc13209cafe8fb3bb5fbd929ec9f11a39e8ac041000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000638bc098000000000000000000000000ef42d748e09a2d9ef89238c053ce0b6f00236210000000000000000000000000000000000000000000000000000000000000014000000000000000000000000000000000000000000000000000000000000000248a0dac4a000000000000000000000000c69fb085481bc8c4bff99b924076656305d9a25d00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000850726f706f736564000000000000000000000000000000000000000000000000",
    "event": "CallExecuted",
    "eventSignature": "CallExecuted(bytes32,uint256,address,uint256,bytes,address,string)",
    "logIndex": 6,
    "removed": false,
    "topics": Array [
      "0x21f5c7d6a15760b2a1e87a9b8cff0ad5e6423cd8f41e04f2196934b1d3859de6",
      "0xe9bfab585a583ae5cc2395f40bfb61b5c93dee4356079f1c642d7157be8300c8",
      "0x0000000000000000000000000000000000000000000000000000000000000000",
    ],
    "transactionHash": "0xdb286638e5dc3a62c2c814fe61016ad0e49349bbba0fcce2b36640a42b70cac3",
    "transactionIndex": 2,
  },
  Object {
    "address": "0xC4611233328559Dd1849947bC1e2b3B6b8714fa9",
    "args": Array [
      "0xe9bfab585a583ae5cc2395f40bfb61b5c93dee4356079f1c642d7157be8300c8",
      Object {
        "hex": "0x01",
        "type": "BigNumber",
      },
      "0xe774D1FB3133b037AA17D39165b8F45f444f632d",
      Object {
        "hex": "0x00",
        "type": "BigNumber",
      },
      "0xc7b9d530000000000000000000000000e8ea1d8b3a5a4cec7e94ae330ff18e82b5d22fa6",
      "0x0000000000000000000000000000000000000000000000000000000000000000",
      "Executed",
    ],
    "blockHash": "0xb48bea6fe4f87cd4323437a9700529db1a8ab5f0376835e46a403f5dfe16806f",
    "blockNumber": 34941481,
    "data": "0x000000000000000000000000e774d1fb3133b037aa17d39165b8f45f444f632d000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000638bc098000000000000000000000000ef42d748e09a2d9ef89238c053ce0b6f0023621000000000000000000000000000000000000000000000000000000000000001400000000000000000000000000000000000000000000000000000000000000024c7b9d530000000000000000000000000e8ea1d8b3a5a4cec7e94ae330ff18e82b5d22fa600000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000850726f706f736564000000000000000000000000000000000000000000000000",
    "event": "CallExecuted",
    "eventSignature": "CallExecuted(bytes32,uint256,address,uint256,bytes,address,string)",
    "logIndex": 7,
    "removed": false,
    "topics": Array [
      "0x21f5c7d6a15760b2a1e87a9b8cff0ad5e6423cd8f41e04f2196934b1d3859de6",
      "0xe9bfab585a583ae5cc2395f40bfb61b5c93dee4356079f1c642d7157be8300c8",
      "0x0000000000000000000000000000000000000000000000000000000000000001",
    ],
    "transactionHash": "0xdb286638e5dc3a62c2c814fe61016ad0e49349bbba0fcce2b36640a42b70cac3",
    "transactionIndex": 2,
  },
  Object {
    "address": "0xC4611233328559Dd1849947bC1e2b3B6b8714fa9",
    "args": Array [
      "0xe9bfab585a583ae5cc2395f40bfb61b5c93dee4356079f1c642d7157be8300c8",
      "0xef42D748e09A2d9eF89238c053CE0B6f00236210",
      "Was kind bad",
      "Cencelled",
    ],
    "blockHash": "0xb48bea6fe4f87cd4323437a9700529db1a8ab5f0376835e46a403f5dfe16806f",
    "blockNumber": 34941482,
    "data": "0x000000000000000000000000fc13209cafe8fb3bb5fbd929ec9f11a39e8ac041000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000638bc098000000000000000000000000ef42d748e09a2d9ef89238c053ce0b6f00236210000000000000000000000000000000000000000000000000000000000000014000000000000000000000000000000000000000000000000000000000000000248a0dac4a000000000000000000000000c69fb085481bc8c4bff99b924076656305d9a25d00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000850726f706f736564000000000000000000000000000000000000000000000000",
    "event": "Cancelled",
    "eventSignature": "Cancelled(bytes32,address,string,string)",
    "logIndex": 6,
    "removed": false,
    "topics": Array [
      "0x21f5c7d6a15760b2a1e87a9b8cff0ad5e6423cd8f41e04f2196934b1d3859de6",
      "0xe9bfab585a583ae5cc2395f40bfb61b5c93dee4356079f1c642d7157be8300c8",
      "0x0000000000000000000000000000000000000000000000000000000000000000",
    ],
    "transactionHash": "0xdb286638e5dc3a62c2c814fe61016ad0e49349bbba0fcce2b36640a42b70cac3",
    "transactionIndex": 2,
  },
  Object {
    "address": "0xC4611233328559Dd1849947bC1e2b3B6b8714fa9",
    "args": Array [
      "0xe9bfab585a583ae5cc2395f40bfb61b5c93dee4356079f1c642d7157be8300c8",
      "0xFc13209cAfE8fb3bb5fbD929eC9F11a39e8Ac041",
      "For lots of reasons",
      "Rejected",
    ],
    "blockHash": "0xb48bea6fe4f87cd4323437a9700529db1a8ab5f0376835e46a403f5dfe16806f",
    "blockNumber": 34941482,
    "data": "0x000000000000000000000000fc13209cafe8fb3bb5fbd929ec9f11a39e8ac041000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000638bc098000000000000000000000000ef42d748e09a2d9ef89238c053ce0b6f00236210000000000000000000000000000000000000000000000000000000000000014000000000000000000000000000000000000000000000000000000000000000248a0dac4a000000000000000000000000c69fb085481bc8c4bff99b924076656305d9a25d00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000850726f706f736564000000000000000000000000000000000000000000000000",
    "event": "Rejected",
    "eventSignature": "Rejected(bytes32,address,string,string)",
    "logIndex": 6,
    "removed": false,
    "topics": Array [
      "0x21f5c7d6a15760b2a1e87a9b8cff0ad5e6423cd8f41e04f2196934b1d3859de6",
      "0xe9bfab585a583ae5cc2395f40bfb61b5c93dee4356079f1c642d7157be8300c8",
      "0x0000000000000000000000000000000000000000000000000000000000000000",
    ],
    "transactionHash": "0xdb286638e5dc3a62c2c814fe61016ad0e49349bbba0fcce2b36640a42b70cac3",
    "transactionIndex": 2,
  },
  Object {
    "address": "0xC4611233328559Dd1849947bC1e2b3B6b8714fa9",
    "args": Array [
      "0xe9bfab585a583ae5cc2395f40bfb61b5c93dee4356079f1c642d7157be8300c8",
      "0xef42D748e09A2d9eF89238c053CE0B6f00236210",
      "Was kind bad",
      "Cencelled",
    ],
    "blockHash": "0xb48bea6fe4f87cd4323437a9700529db1a8ab5f0376835e46a403f5dfe16806f",
    "blockNumber": 34941483,
    "data": "0x000000000000000000000000e774d1fb3133b037aa17d39165b8f45f444f632d000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000638bc098000000000000000000000000ef42d748e09a2d9ef89238c053ce0b6f0023621000000000000000000000000000000000000000000000000000000000000001400000000000000000000000000000000000000000000000000000000000000024c7b9d530000000000000000000000000e8ea1d8b3a5a4cec7e94ae330ff18e82b5d22fa600000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000850726f706f736564000000000000000000000000000000000000000000000000",
    "event": "Cancelled",
    "eventSignature": "Cancelled(bytes32,address,string,string)",
    "logIndex": 7,
    "removed": false,
    "topics": Array [
      "0x21f5c7d6a15760b2a1e87a9b8cff0ad5e6423cd8f41e04f2196934b1d3859de6",
      "0xe9bfab585a583ae5cc2395f40bfb61b5c93dee4356079f1c642d7157be8300c8",
      "0x0000000000000000000000000000000000000000000000000000000000000001",
    ],
    "transactionHash": "0xdb286638e5dc3a62c2c814fe61016ad0e49349bbba0fcce2b36640a42b70cac3",
    "transactionIndex": 2,
  },
  Object {
    "address": "0xC4611233328559Dd1849947bC1e2b3B6b8714fa9",
    "args": Array [
      "0xe9bfab585a583ae5cc2395f40bfb61b5c93dee4356079f1c642d7157be8300c8",
      "0xFc13209cAfE8fb3bb5fbD929eC9F11a39e8Ac041",
      "For lots of reasons",
      "Rejected",
    ],
    "blockHash": "0xb48bea6fe4f87cd4323437a9700529db1a8ab5f0376835e46a403f5dfe16806f",
    "blockNumber": 34941483,
    "data": "0x000000000000000000000000e774d1fb3133b037aa17d39165b8f45f444f632d000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000638bc098000000000000000000000000ef42d748e09a2d9ef89238c053ce0b6f0023621000000000000000000000000000000000000000000000000000000000000001400000000000000000000000000000000000000000000000000000000000000024c7b9d530000000000000000000000000e8ea1d8b3a5a4cec7e94ae330ff18e82b5d22fa600000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000850726f706f736564000000000000000000000000000000000000000000000000",
    "event": "Rejected",
    "eventSignature": "Rejected(bytes32,address,string,string)",
    "logIndex": 7,
    "removed": false,
    "topics": Array [
      "0x21f5c7d6a15760b2a1e87a9b8cff0ad5e6423cd8f41e04f2196934b1d3859de6",
      "0xe9bfab585a583ae5cc2395f40bfb61b5c93dee4356079f1c642d7157be8300c8",
      "0x0000000000000000000000000000000000000000000000000000000000000001",
    ],
    "transactionHash": "0xdb286638e5dc3a62c2c814fe61016ad0e49349bbba0fcce2b36640a42b70cac3",
    "transactionIndex": 2,
  },
]
`;

exports[`governance.service loadScheduledProposals should return scheduled proposals 1`] = `
Array [
  Object {
    "address": "0xC4611233328559Dd1849947bC1e2b3B6b8714fa9",
    "args": Array [
      "0xe9bfab585a583ae5cc2395f40bfb61b5c93dee4356079f1c642d7157be8300c8",
      Object {
        "hex": "0x00",
        "type": "BigNumber",
      },
      "0xFc13209cAfE8fb3bb5fbD929eC9F11a39e8Ac041",
      Object {
        "hex": "0x00",
        "type": "BigNumber",
      },
      "0x8a0dac4a000000000000000000000000c69fb085481bc8c4bff99b924076656305d9a25d",
      "0x0000000000000000000000000000000000000000000000000000000000000000",
      Object {
        "hex": "0x638bc098",
        "type": "BigNumber",
      },
      "0xef42D748e09A2d9eF89238c053CE0B6f00236210",
      "Lets invest all treasury funds into Luna tokens",
      "Proposed",
    ],
    "blockHash": "0xb48bea6fe4f87cd4323437a9700529db1a8ab5f0376835e46a403f5dfe16806f",
    "blockNumber": 34941487,
    "data": "0x000000000000000000000000fc13209cafe8fb3bb5fbd929ec9f11a39e8ac041000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000638bc098000000000000000000000000ef42d748e09a2d9ef89238c053ce0b6f00236210000000000000000000000000000000000000000000000000000000000000014000000000000000000000000000000000000000000000000000000000000000248a0dac4a000000000000000000000000c69fb085481bc8c4bff99b924076656305d9a25d00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000850726f706f736564000000000000000000000000000000000000000000000000",
    "event": "CallScheduled",
    "eventSignature": "CallScheduled(bytes32,uint256,address,uint256,bytes,bytes32,uint256,address,string,string)",
    "logIndex": 6,
    "removed": false,
    "topics": Array [
      "0x21f5c7d6a15760b2a1e87a9b8cff0ad5e6423cd8f41e04f2196934b1d3859de6",
      "0xe9bfab585a583ae5cc2395f40bfb61b5c93dee4356079f1c642d7157be8300c8",
      "0x0000000000000000000000000000000000000000000000000000000000000000",
    ],
    "transactionHash": "0xdb286638e5dc3a62c2c814fe61016ad0e49349bbba0fcce2b36640a42b70cac3",
    "transactionIndex": 2,
  },
  Object {
    "address": "0xC4611233328559Dd1849947bC1e2b3B6b8714fa9",
    "args": Array [
      "0xe9bfab585a583ae5cc2395f40bfb61b5c93dee4356079f1c642d7157be8300c8",
      Object {
        "hex": "0x01",
        "type": "BigNumber",
      },
      "0xe774D1FB3133b037AA17D39165b8F45f444f632d",
      Object {
        "hex": "0x00",
        "type": "BigNumber",
      },
      "0xc7b9d530000000000000000000000000e8ea1d8b3a5a4cec7e94ae330ff18e82b5d22fa6",
      "0x0000000000000000000000000000000000000000000000000000000000000000",
      Object {
        "hex": "0x638bc098",
        "type": "BigNumber",
      },
      "0xef42D748e09A2d9eF89238c053CE0B6f00236210",
      "Lets invest all treasury funds into Luna tokens",
      "Proposed",
    ],
    "blockHash": "0xb48bea6fe4f87cd4323437a9700529db1a8ab5f0376835e46a403f5dfe16806f",
    "blockNumber": 34941487,
    "data": "0x000000000000000000000000e774d1fb3133b037aa17d39165b8f45f444f632d000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000638bc098000000000000000000000000ef42d748e09a2d9ef89238c053ce0b6f0023621000000000000000000000000000000000000000000000000000000000000001400000000000000000000000000000000000000000000000000000000000000024c7b9d530000000000000000000000000e8ea1d8b3a5a4cec7e94ae330ff18e82b5d22fa600000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000850726f706f736564000000000000000000000000000000000000000000000000",
    "event": "CallScheduled",
    "eventSignature": "CallScheduled(bytes32,uint256,address,uint256,bytes,bytes32,uint256,address,string)",
    "logIndex": 7,
    "removed": false,
    "topics": Array [
      "0x21f5c7d6a15760b2a1e87a9b8cff0ad5e6423cd8f41e04f2196934b1d3859de6",
      "0xe9bfab585a583ae5cc2395f40bfb61b5c93dee4356079f1c642d7157be8300c8",
      "0x0000000000000000000000000000000000000000000000000000000000000001",
    ],
    "transactionHash": "0xdb286638e5dc3a62c2c814fe61016ad0e49349bbba0fcce2b36640a42b70cac3",
    "transactionIndex": 2,
  },
]
`;

exports[`governance.service timelock controller address getter should return the timeLock address 1`] = `"0xd2ac6abdd6acbdfab7228f5d4983de2f643d5735"`;
